lib/utils/CHIMETelescope.cpp

```diff
@@ -15,29 +16,72 @@
 #include <vector>    // for vector
 
 
-REGISTER_TELESCOPE(CHIMETelescope, "chime");
+REGISTER_TELESCOPE(CHIMETelescope, "CHIMETelescope");
 
 CHIMETelescope::CHIMETelescope(const kotekan::Config& config, const std::string& path) :
     ICETelescope(config.get<std::string>(path, "log_level")) {
 
     // This is always 1 for CHIME
     _num_freq_per_stream = 1;
 
+    set_sampling_params(800.0, 2048, 2);
+
+    // Get the GPS time, either from the config or fpga_master
+    bool require_gps = config.get_default<bool>(path, "require_gps", true);
+    _query_gps = config.get_default<bool>(path, "query_gps", false);
+    _gps_host = config.get_default<std::string>(path, "gps_host", "10.1.13.1");
+    _gps_port = config.get_default<uint32_t>(path, "gps_port", 54321);
+    _gps_endpoint = config.get_default<std::string>(path, "gps_endpoint", "/get-frame0-time");
+    if (_query_gps)
+        set_gps(_gps_host, _gps_port, _gps_endpoint);
+    if (!gps_enabled)
+        set_gps(config);
+
     _require_frequency_map = config.get_default<bool>(path, "require_frequency_map", false);
     _allow_default_frequency_map =
         config.get_default<bool>(path, "allow_default_frequency_map", true);
 
-    set_sampling_params(800.0, 2048, 2);
-    set_gps(config, path);
-    set_frequency_map(config, "fpga_frequency_map");
+    // Get the frequency map, either from config or fpga_master
+    _query_frequency_map = config.get_default<bool>(path, "query_frequency_map", false);
+    _frequency_map_host = config.get_default<std::string>(path, "frequency_map_host", "10.1.13.1");
+    _frequency_map_port = config.get_default<uint32_t>(path, "frequency_map_port", 54321);
+    _frequency_map_endpoint =
+        config.get_default<std::string>(path, "frequency_map_endpoint", "/get-frequency-map");
+
+    nlohmann::json fpga_freq_map_json;
+    if (_query_frequency_map)
+        fpga_freq_map_json =
+            get_frequency_map(_frequency_map_host, _frequency_map_port, _frequency_map_endpoint);
+    else
+        fpga_freq_map_json = config.get_default<nlohmann::json>("/", "fpga_frequency_map", {});
+
+    set_frequency_map(fpga_freq_map_json);
+
+    if (require_gps && !gps_enabled) {
+        throw std::runtime_error("The system requires a GPS time, but none was found.");
+    }
 }
 
-void CHIMETelescope::set_frequency_map(const kotekan::Config& config, const std::string& path) {
-    nlohmann::json fpga_freq_map_json = config.get_default<nlohmann::json>("/", path, {});
+nlohmann::json CHIMETelescope::get_frequency_map(const std::string& host, const uint32_t port,
```

Comment:
I would call this method something like `fetch_frequency_map`. As `get_/set_` methods are a common pattern, you might believe that this one returns the frequency map currently set in kotekan, but it doesn't.

Reply:
That's a very good point; I've fixed that now.   Thanks for the review!